demos/coreMQTT/mqtt_demo_keep_alive.c

```diff
@@ -0,0 +1,1177 @@
+        {
+            /* Publish messages with QOS0, send and process keep-alive messages. */
+            for( ulPublishCount = 0; ulPublishCount < ulMaxPublishCount; ulPublishCount++ )
+            {
+                LogInfo( ( "Publish to the MQTT topic %s.", mqttexampleTOPIC ) );
+                xDemoStatus = prvMQTTPublishToTopic( &xMQTTContext );
+
+                if( xDemoStatus == pdPASS )
+                {
+                    /* Process incoming publish echo, since application subscribed
+                     * to the same topic the broker will send the publish message
+                     * back to the application. */
+                    LogInfo( ( "Attempt to receive publish message from broker." ) );
+                    xMQTTStatus = MQTT_ReceiveLoop( &xMQTTContext,
+                                                    mqttexampleRECEIVE_LOOP_TIMEOUT_MS );
+
+                    if( xMQTTStatus != MQTTSuccess )
+                    {
+                        LogError( ( "MQTT_ReceiveLoop() failed with status %s.",
+                                    MQTT_Status_strerror( xMQTTStatus ) ) );
+                        xDemoStatus = pdFAIL;
+                    }
+                }
+
+                /* Leave Connection Idle for some time. */
+                LogInfo( ( "Keeping Connection Idle..." ) );
+                vTaskDelay( mqttexampleDELAY_BETWEEN_PUBLISHES );
+            }
+        }
+
+        /********************** Unsubscribe from the topic. ***********************/
+
+        if( xDemoStatus == pdPASS )
+        {
+            LogInfo( ( "Unsubscribe from the MQTT topic %s.", mqttexampleTOPIC ) );
+            xDemoStatus = prvMQTTUnsubscribeFromTopic( &xMQTTContext );
+        }
+
+        /* Process Incoming packet from the broker. */
+        if( xDemoStatus == pdPASS )
+        {
+            xMQTTStatus = MQTT_ReceiveLoop( &xMQTTContext, mqttexampleRECEIVE_LOOP_TIMEOUT_MS );
+
+            if( xMQTTStatus != MQTTSuccess )
+            {
+                LogError( ( "MQTT_ReceiveLoop() failed with status %s.",
+                            MQTT_Status_strerror( xMQTTStatus ) ) );
+                xDemoStatus = pdFAIL;
+            }
+        }
+
+        /**************************** Disconnect. *****************************/
+
+        /* Whether the demo operations preceeding are successful or not, we
+         * always disconnect an open connection to free up system resources. */
+        if( xIsConnectionEstablished == pdTRUE )
+        {
+            /* Send an MQTT Disconnect packet over the already connected TCP socket.
+             * There is no corresponding response for the disconnect packet. After
+             * sending the disconnect, the client must close the network connection. */
+            LogInfo( ( "Disconnecting the MQTT connection with %s.",
+                       democonfigMQTT_BROKER_ENDPOINT ) );
+            xMQTTStatus = MQTT_Disconnect( &xMQTTContext );
```

Comment:
Should `MQTT_Disconnect` be called if `MQTT_Connect` had failed after a successful establishment of the TCP connection?

Reply:
It's not unsafe.

Reply:
My main point is that the error message about `MQTT_Disconnect` failure would be misleading when the `MQTT_Connect` fail had itself not succeeded

Reply:
I will turn it into a warn.

Reply:
In the case where everything passed except MQTT_Disconnect(), we still want to consider that a failing demo. That is amiss in the mutual auth demo.